Add function to generate level 3 data from order

v18 opened the pull request:
Fixes #860.

This PR sends level3 data in the following functions:

1. update_existing_intent
   - Can be tested by checking out with a failing credit card (e.g. `4000000000000069`) and then checking out with a working card.
2. confirm_intent
   - Can be tested by checking out normally.
3. create_and_confirm_intent_for_off_session
   - Can be tested by renewal a subscription (Edit Subscription screen > Subscription actions > Renew subscription).
4. prepare_intent_for_order_pay_page
   - Can be tested by purchasing a subscription using card that requires 3DS authentication (e.g. `4000 0027 6000 3184`), then renewing the subscription, and then using the authorize link sent to the customer to authenticate the subscription.
5. capture_payment
   - Can be tested by unchecking the option "Capture charge immediately" in the Stripe settings page at `wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe`

Stripe doc: https://stripe.com/docs/level3

This is what the level 3 data looks like:

```
level3: {
	'merchant_reference', // Order ID
	'shipping_address_zip', // The customer’s U.S. shipping ZIP code.
	'shipping_from_zip', // The merchant’s U.S. shipping ZIP code.
	'shipping_amount', // The shipping cost, in cents, as a non-negative integer.
	'line_items': [ // Object per item in order
		{
			'product_code', // Up to 12 characters that uniquely identify the product.
				'product_description', // Up to 26 characters long describing the product.
				'unit_cost', // Cost of the product, in cents, as a non-negative integer.
				'quantity', // The number of items of this type sold, as a non-negative integer.
				'tax_amount', // The amount of tax this item had added to it, in cents, as a non-negative integer.
				'discount_amount', // The amount an item was discounted—if there was a sale,for example, as a non-negative integer.
		},
	]
}
```

You can test it with an account where level3 is allowed by installing the WooCommerce Services plugin, going through the WC setup wizard, and option to create a new Stripe account (using a new email address that hasn't been used with Stripe before). 

The request with level3 data can fail for a couple of reasons, which are handled by retrying the request without level3 data:

1. If it fails because level3 data is not permitted, level3 data will not be added to all subsequent requests won't using a transient.
2. If it fails because the level3 data we send has an incorrect sum, the request will be tried again without the data, and the error will be logged.

Some notes which have been verified with Stripe:

- Data will be sent for all merchants and customers, not only those in the US. This has been verified with Stripe.
- We can send non-US postal codes in the level3 data object
- The level 3 data should be sent in the currency that's used for the order
- The product code can be the product ID
- Level 3 data should be sent for both digital and shippable products, and for $0 products
- The billing details should be sent using source objects, which we already do.
- We are not going to supply the `customer_reference` parameter (paJDYF-6e-p2#comment-586), which is why it's missing (vs the docs).

Here are some things that are not fully resolved as part of this PR:

- What happens for stores with multiple shipping zipcodes? There isn't a way in WooCommerce to identify which originating zipcode will be used as part of an order. I don't think it should be fixed as part of this PR.
- Unit testing with taxes and shipping

allendav commented:
> Level 3 data should be sent under these circumstances:
> When the store is located in the US
> When the purchase is e-commerce and not retail based. Currently all of Stripe's payments are assumed to be e-commerce.

Please document how we came to these conclusions - I wasn't aware of the scope reduction here

allendav commented:
> What happens for stores with multiple shipping zipcodes? There isn't a way in WooCommerce to identify which originating zipcode will be used as part of an order.

Please open a separate issue for that and link it to this issue. That seems like an edge case to me? What do you think?

allendav commented:
> Some more questions with Stripe are unresolved and being discussed in p1580738023019900-slack-woo-stripe-external

1) Should we be sending level 3 data for non-US merchants?
2) Should we be sending level 3 data for non-US customers?
3) Should we be sending level 3 data for non-USD currencies?
4) For the product code - can I use a product code that’s unique for the store? For example the product ID?
5) Should we be sending level 3 data only for products that will be shipped, or all products (e.g. electronic downloads)?
6) Should we be sending information about free / $0 products in the line items array if they are included as part of the order?
7) Discounts vs sales. For sale items in WooCommerce stores, we display the sale item up front, e.g. Full price $17, sale price $10. Customers can add further discounts using coupon codes. Can we send the unit cost as $10 for this sort of sale item, where no further discount is applied?
8) I see that the documentation mentions “You must provide the customer billing details—name, address, city, state, ZIP code—associated with the card.“, but fields for billing details are not in the documented level 3 object. Should we be sending customer billing details?

Quinn/Stripe replied:

1 not necessary but it's totally fine if you do for simplicity
2 yes
3 yes
4 yes that should work great
5 all products
6 yes
7 yes
8 yes

v18 commented:
Ah, I need to update how I test for array equality in the tests.

RadoslavGeorgiev commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1055,6 +1055,49 @@ public function generate_create_intent_request( $order, $prepared_source ) {
+			$product_id          = $item->get_variation_id()
+				? $item->get_variation_id()
+				: $item->get_product_id();
+			$product             = wc_get_product( $product_id );
```
`WC_Order_Item_Product::get_product()` does something very similar (check for a variation, load the product, etc.). Any reason not to use the method?

v18 replied:
To follow up here, it's using the item directly now 👍

v18 commented:
@RadoslavGeorgiev hey - I updated this branch with these changes:

- Rebased on `master`
- Explicitly removed support for WC version < 3.0 since it's not possible to get the shipping location zipcode and/or the store base zipcode.
- Updated the product name to use `$item->name`
- For shipping total, added the shipping tax to the total amount, since it's not part of `get_shipping_total()`.
- Removed checking for mention of "level3" in the error message since we can't rely on it for all errors related to level3. Instead, we'll be logging all invalid request types at that stage, and trying the request again without level3 data.
- Removed `print_r( $level3, true ) . PHP_EOL` from the logging since $level3 is not defined.

RadoslavGeorgiev commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1150,7 +1150,9 @@ public function update_existing_intent( $intent, $order, $prepared_source ) {
-		$level3_data = $this->get_level3_data_from_order( $order, get_option( 'woocommerce_store_postcode' ) );
+		$level3_data = WC_Stripe_Helper::is_wc_lt( '3.0' )
+			? array()
+			: $this->get_level3_data_from_order( $order, get_option( 'woocommerce_store_postcode' ) );
```
Why are you individually checking for pre-3.0 on every `get_level3_data_from_order` call? Seems like you can move this logic within the method and only have it once.

v18 replied:
Yeh, I thought it was better to be explicit here but I can see that it's noisy. Moved the check to `get_level3_data_from_order` 👌

RadoslavGeorgiev commented on includes/class-wc-stripe-api.php:
```diff
@@ -194,6 +194,16 @@ public static function retrieve( $api ) {
 	 */
 	public static function request_with_level3_data( $request, $api, $level3_data, $order ) {
+		// Do not add level3 data for WC versions less than 3.0,
+		// since the store zip code is not available.
```
Let's move this comment to `$level3_data = WC_Stripe_Helper::is_wc_lt( '3.0' ) ? :` calls.

v18 replied:
👌I have a comment in both places now.

v18 commented:
@RadoslavGeorgiev - I updated the PR with your suggestions. Also added an expiration to the timeout so that we can attempt to send level3 data again for each site every 3 months.

Here's a screenshot of a request with level3 data:

<img width="1101" alt="Screen Shot 2020-02-11 at 2 34 14 PM" src="https://user-images.githubusercontent.com/11487924/74213240-b10a1c00-4cdb-11ea-9e15-be08bcebf63d.png">

RadoslavGeorgiev commented:
In my last few commits I thought It'd add support for pre-3.0 (since we're still supporting it, even though that'll stop soon), but that would require too many `WC_Stripe_Helper::is_wc_lt( '3.0' )` to be worth it.

Additionally, I included the new `is_valid_us_zip_code` method to check whether postal codes have the correct format for the US, because I was having issues with Stripe validation when using non-US stores.

reykjalin commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1055,6 +1055,67 @@ public function generate_create_intent_request( $order, $prepared_source ) {
+		// (because it's not possible to get a store postcode before WC 3.0).
+		if ( WC_Stripe_Helper::is_wc_lt( '3.0' ) ) {
+			return array();
+		}
```
>Return an empty array if there's no store postcode

This seems to not be true anymore?

RadoslavGeorgiev replied:
Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1130/commits/e38959829e18a9364271ef2e477582c1c9ea0c0d

reykjalin commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1055,6 +1055,67 @@ public function generate_create_intent_request( $order, $prepared_source ) {
+		$stripe_line_items = array_map(function( $item ) use ( $currency ) {
+			$product_id          = $item->get_variation_id()
+				? $item->get_variation_id()
+				: $item->get_product_id();
```
Do we need to limit the `ID` here to 12 characters similar to what we do for `$product_description`, or are we certain the `ID` will always be 12 characters or less?

RadoslavGeorgiev replied:
The ID in this case is an integer, coming from the product/post ID. The DB column is a BIGINT(20), however I don't expect any issues with this. If someone's store has 999999999999+ posts, I think they have bigger problems to worry about.

reykjalin replied:
hahaha sounds about right 😛

reykjalin commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1055,6 +1055,67 @@ public function generate_create_intent_request( $order, $prepared_source ) {
+			$quantity            = $item->get_quantity();
+			$unit_cost           = WC_Stripe_Helper::get_stripe_amount( ( $item->get_subtotal() / $quantity ), $currency );
+			$tax_amount          = WC_Stripe_Helper::get_stripe_amount( $item->get_total_tax(), $currency );
+			$discount_amount     = WC_Stripe_Helper::get_stripe_amount( $item->get_subtotal() - $item->get_total(), $currency );
```
Do we need to verify that these numbers (except `$tax_amount`) are non-negative integers, or is that guaranteed?

RadoslavGeorgiev replied:
`WC_Stripe_Helper::get_stripe_amount` always uses `absint()` before returning, so yes.

reykjalin commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1055,6 +1055,67 @@ public function generate_create_intent_request( $order, $prepared_source ) {
+		$store_postcode = get_option( 'woocommerce_store_postcode' );
+		if ( $this->is_valid_us_zip_code( $store_postcode ) ) {
+			$level3_data['shipping_from_zip'] = $store_postcode;
+		}
```
Is there no error handling we need to do here? What happens if the ZIP codes aren't valid? What happens if one is invalid, and the other one isn't?

RadoslavGeorgiev replied:
> What happens if the ZIP codes aren't valid?

AFAICT Stripe does not validate the ZIP codes based on any other factor except format, so the validation in `is_valid_us_zip_code()` method must be sufficient.

> What happens if one is invalid, and the other one isn't?

Both fields are optional. The point here is to provide as much data as possible, the combination does not matter.

reykjalin commented on includes/class-wc-stripe-api.php:
```diff
@@ -176,4 +176,97 @@ public static function retrieve( $api ) {
+				$request,
+				$api
+			);
+		}
```
I think it would be better to just `return` instead of setting the value of `$result` since AFAICT you can't enter the next `if` statement anyway.

RadoslavGeorgiev replied:
Why not? The purpose of the `if` statement is to log `invalid_request_error` requests in order to be able to debug them. If the response from the API contains such an error, this can (and does) happen.

reykjalin replied:
My expectation was that `is_level_3data_incorrect` (the check before the next `if` statement) would never be true because `$request['level3']` is unset before sending the data in **this** `if` statement, i.e. how can the level 3 data be incorrect if the level 3 data wasn't sent (because level 3 data isn't supported).

If we can still enter the second `if` statement after having sent a second request, I'd suggest naming the check something other than `$is_level_3data_incorrect`, e.g. just `$is_request_data_incorrect` or something along those lines.

In essence I think the variable name of `$is_level_3data_incorrect` caused some confusion here for me 🙂.

RadoslavGeorgiev replied:
Although I did not rename the variable, I see how it could be confusing, so I restructured it here: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1130/commits/880fddc08e0a5cb56d2aad74b460b2968fae56c0. What do you think?

reykjalin replied:
Perfect 👌

reykjalin commented on includes/class-wc-stripe-api.php:
```diff
@@ -176,4 +176,97 @@ public static function retrieve( $api ) {
+				$request,
+				$api
+			);
+		}
```
I think it would be better to just `return` instead of setting the value of `$result` since AFAICT you can't enter the next `if` statement anyway.

RadoslavGeorgiev replied:
Switched to `return` in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1130/commits/e38959829e18a9364271ef2e477582c1c9ea0c0d

nirkons commented:
Chiming in, also having this issue unfortunately
